### ISSUE TITLE
__NSCFLocalDownloadTask instance deallocated while KVO still registered with it.

### ISSUE THREAD

**Issue by HorsieSoftware** [AFNetworking/AFNetworking#3644]
```
An instance 0x130414630 of class __NSCFLocalDownloadTask was deallocated while key value observers were still registered with it. Current observation info: NSKeyValueObservationInfo 0x17564ec70 (
<NSKeyValueObservance 0x1746daef0: Observer: 0x174a9e820, Key path: countOfBytesReceived, Options: New: YES, Old: NO, Prior: NO Context: 0x0, Property: 0x170258bd0
<NSKeyValueObservance 0x1746da470: Observer: 0x174a9e820, Key path: countOfBytesExpectedToReceive, Options: New: YES, Old: NO, Prior: NO Context: 0x0, Property: 0x170254d00
<NSKeyValueObservance 0x1746da400: Observer: 0x174a9e820, Key path: countOfBytesSent, Options: New: YES, Old: NO, Prior: NO Context: 0x0, Property: 0x1744449e0
<NSKeyValueObservance 0x1746dae10: Observer: 0x174a9e820, Key path: countOfBytesExpectedToSend, Options: New: YES, Old: NO, Prior: NO Context: 0x0, Property: 0x174444b90
)
0   CoreFoundation       0x1857282d8 __exceptionPreprocess + 132
1   libobjc.A.dylib      0x196f540e4 objc_exception_throw + 60
2   CoreFoundation       0x185728218 -[NSException initWithCoder:] + 0
3   Foundation           0x1865ad810 NSKVODeallocate + 340
4   libsystem_blocks.dylib 0x1975fd910 _Block_release + 256
5   Foundation           0x186561dbc -[NSBlockOperation dealloc] + 68
6   Foundation           0x18661201c __NSOQSchedule_f + 472
7   libdispatch.dylib    0x1975a5954 _dispatch_client_callout + 16
8   libdispatch.dylib    0x1975b00a4 _dispatch_queue_drain + 1448
9   libdispatch.dylib    0x1975a8a5c _dispatch_queue_invoke + 132
10  libdispatch.dylib    0x1975b2318 _dispatch_root_queue_drain + 720
11  libdispatch.dylib    0x1975b3c4c _dispatch_worker_thread3 + 108
12  libsystem_pthread.dylib 0x19778522c _pthread_wqthread + 816
13  libsystem_pthread.dylib 0x197784ef0 start_wqthread + 4
}
```
This appears related to https://github.com/AFNetworking/AFNetworking/issues/2053, however that was version <= 2.3.1 and the issue was closed in 2014. The only more recent reference to this issue I can find in AFNetworking can be found by searching "__NSCFLocalDownloadTask" in here:
https://travis-ci.org/AFNetworking/AFNetworking/jobs/96065600

I am seeing this in the wild using AFNetworking version 3.0. I have had consistent reports - users on iPhone 6 plus, iOS 8.3. It happens every time with this combination, but disappears on upgrade to iOS 9.3.4. This is the helper method I use to download files with AFNetworking and should be the point of contact with your code:

```
+ (void) downloadFileFromURL:(NSString *)stringURL
                  toFilePath:(NSString *)filePath
                   withStart:(void (^)(void))startBlock
                    progress:(BOOL (^)(NSURLSessionTaskState taskState, CGFloat progress))progressBlock
                  completion:(void (^)(NSURL *filePath))completionBlock
                     failure:(void (^)(NSError *error))failureBlock
{
    if (!stringURL || !filePath) return;
    NSURLSessionConfiguration *sessionConfig = [NSURLSessionConfiguration defaultSessionConfiguration];
    AFURLSessionManager *sessionManager = [[AFURLSessionManager alloc] initWithSessionConfiguration:sessionConfig];
    
    NSURL *url = [NSURL URLWithString:stringURL];
    NSURLRequest *request = [NSURLRequest requestWithURL:url];
    
    if (progressBlock)
    {
        __weak AFURLSessionManager *weakSessionManager = sessionManager; // weak reference to avoid retain cycle
        [sessionManager setDownloadTaskDidWriteDataBlock:^(NSURLSession * _Nonnull session, NSURLSessionDownloadTask * _Nonnull task, int64_t bytesWritten, int64_t totalBytesWritten, int64_t totalBytesExpectedToWrite) {
            
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0ul), ^{
                CGFloat progress = (float)totalBytesWritten/(float)totalBytesExpectedToWrite;
                if (progressBlock([task state], progress)) 
                {
                    // Progress block returns YES if a cancellation is requested mid-download
                    [weakSessionManager invalidateSessionCancelingTasks:YES];
                }
            });
        }];
    }
    
    NSURLSessionDownloadTask *downloadTask = [sessionManager downloadTaskWithRequest:request
                                                                            progress:nil
                                                                         destination:^NSURL * _Nonnull(NSURL * _Nonnull targetPath, NSURLResponse * _Nonnull response) {
                                                                             return [NSURL fileURLWithPath:filePath];
                                                                         }
                                                                   completionHandler:^(NSURLResponse * _Nonnull response, NSURL * _Nullable filePath, NSError * _Nullable error) {
                                                                       dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0ul), ^{
                                                                          
                                                                           if (!error && completionBlock) 
                                                                           {
                                                                               completionBlock(filePath);
                                                                           }
                                                                           else if (failureBlock) 
                                                                           {
                                                                               failureBlock(error);
                                                                           }
                                                                       });

                                                                       [sessionManager invalidateSessionCancelingTasks:NO];
                                                                   }];
    if (startBlock) startBlock();
    [downloadTask resume];
}
```

**Comment by laoyao** [comment]
My project have the same problem. Only receive crash log at 8.x system.

**Comment by avaniashev** [comment]
I have the same issue in a simulator 9.2. AFNetworking version 3.1.0

`'NSInternalInconsistencyException', reason: 'An instance 0x7f97c4378360 of class __NSCFLocalDataTask was deallocated while key value observers were still registered with it. Current observation info: <NSKeyValueObservationInfo 0x7f97c4196fd0> (
<NSKeyValueObservance 0x7f97c4159cf0: Observer: 0x7f97c4154d90, Key path: countOfBytesReceived, Options: <New: YES, Old: NO, Prior: NO> Context: 0x0, Property: 0x7f97c1da8210>
<NSKeyValueObservance 0x7f97c43010f0: Observer: 0x7f97c4154d90, Key path: countOfBytesExpectedToReceive, Options: <New: YES, Old: NO, Prior: NO> Context: 0x0, Property: 0x7f97c4151670>
<NSKeyValueObservance 0x7f97c41d51a0: Observer: 0x7f97c4154d90, Key path: countOfBytesSent, Options: <New: YES, Old: NO, Prior: NO> Context: 0x0, Property: 0x7f97c4151c70>
<NSKeyValueObservance 0x7f97c4154b70: Observer: 0x7f97c4154d90, Key path: countOfBytesExpectedToSend, Options: <New: YES, Old: NO, Prior: NO> Context: 0x0, Property: 0x7f97c41523d0>
)'`

**Comment by willonboy** [comment]
My project have the same problem.

**Comment by MagicLinuxRing** [comment]
I have met same problem,hope @mattt can fix this problem

**Comment by TheHmmka** [comment]
The same here!

**Comment by YBthebest** [comment]
Same problem here.

**Comment by kcharwood** [comment]
We've just pushed some improvements to master. Can you try and see if you are still able to reproduce with that?

**Comment by HorsieSoftware** [comment]
@kcharwood Thank you sir, your efforts are much appreciated! We actually moved our deployment target to 9.0, in part to bypass this issue (we are still in beta and this was only reported on iOS 8), but also for other reasons.

Nevertheless I have integrated your update into our codebase and will include it in future releases.  Should we see any occurence on iOS 9 or 10 I will report back here.  For the record, I was never able to reproduce this on test devices on iOS 8 - it was only something reported from the wild.
